Bgrenon partners guide

bgrenon opened the pull request:
Preview:

- https://neon-next-git-bgrenon-partners-guide-neondatabase.vercel.app/docs/guides/partner-intro
- https://neon-next-git-bgrenon-partners-guide-neondatabase.vercel.app/docs/guides/partner-billing

Questions for reviews:
- I am unsure about some of the actions suggested in the last section [Resetting a project after suspend](https://neon-next-git-bgrenon-partners-guide-neondatabase.vercel.app/docs/guides/partner-billing#resetting-a-project-after-suspend). Please advise.
- A question for Atli or someone who has worked with partners and "quotas". Should this document cover these project limits as well:

  - autoscaling_limit_min_cu
  - autoscaling_limit_max_cu
  - suspend_timeout_seconds
  - history_retention_seconds

danieltprice commented on content/docs/guides/partner-intro.md:
```diff
@@ -0,0 +1,44 @@
+
+<DetailIconCards>
+
+<a href="https://neon.tech/partners" description="Discover the benefits of partnering with Neon for serverless Postgres" icon="hourglass">Partner Page</a>
```
```suggestion
<a href="https://neon.tech/partners" description="Discover the benefits of partnering with Neon for serverless Postgres" icon="hourglass">Partner page</a>
```

danieltprice commented on content/docs/guides/partner-intro.md:
```diff
@@ -0,0 +1,44 @@
+
+<DetailIconCards>
+
+<a href="/docs/guides/partner-billing" description="Use the Neon API to configure billing quotas for your customers" icon="setup">Set up usage quotas</a>
```
The text mentions "billing quotas" and the link says "usage quotas". Should we use one or the other?

bgrenon replied:
Changed to: 

Use the Neon API to configure consumption quotas for your customers.

Set up usage quotas

Question: are "consumption" and "usage" synonymous enough it's a benefit to use both terms here, short in the title, longer in description?

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+* `written_data_bytes`
+* `data_transfer_bytes` 
+
+These consumption metrics represent total cumulated usage across all branches and computes in a given project, accrued so far in a given monthly billing period. They are refreshed on a set day every month, on whichever date your new billing period starts. 
```
Is it any date or is it the first day of the month that a new billing period starts? We'll have to check. I always thought it was the first of the month, but I could be wrong.

bgrenon replied:
Revised to: 

"Metrics are refreshed on the first day of the following month."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+
+The `quota` object includes the array of parameters used to set threshold limits. Their names generally match their corresponding metric:
+
+* `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated before the one-month refresh date.
```
Does active time include the "auto-suspend delay" period as well?

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+
+The `quota` object includes the array of parameters used to set threshold limits. Their names generally match their corresponding metric:
+
+* `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated before the one-month refresh date.
```
Active time might include the auto-suspend delay period as well.

bgrenon replied:
Sounds like no suspended time is excluded. From https://neon.tech/docs/introduction/billing#active-time

"...excludes time when computes are in an Idle state due to [auto-suspension](https://neon.tech/docs/reference/glossary#auto-suspend-compute) (scale-to-zero)."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+The `quota` object includes the array of parameters used to set threshold limits. Their names generally match their corresponding metric:
+
+* `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated before the one-month refresh date.
+* `compute_time_seconds` &#8212; Sets the maximum amount of CPU seconds allowed in total across all of a project's compute endpoints. This differs from `active_time_seconds` in that it only counts the time the CPU spends executing a process. It excludes time spent waiting for external resources, I/O operations, or time spent in a blocked or idle state.
```
I am not sure about "This differs from `active_time_seconds` in that it only counts the time the CPU spends executing a process. It excludes time spent waiting for external resources, I/O operations, or time spent in a blocked or idle state."

The way I understand it is this, but maybe I am wrong:

If a compute with 4 vCPUs runs for 1 second, active_time_seconds=1.
If a compute with 4 vCPUs runs for 1 second, compute_time_seconds=4.

So, the larger your computes, the faster you'll burn through compute_time_seconds.

bgrenon replied:
I've revised it to:

Sets the maximum amount of CPU seconds allowed in total across all of a project's compute endpoints, including any endpoints deleted during the current billing period. Note that the larger your compute endpoints, the faster you'll consume through `compute_time_seconds`. For example, a compute instance with 0.25 vCPU uses 0.25 seconds of compute time for every second it's active. If it has 4 vCPUs, it uses 4 seconds of compute time in one active second.
   | vCPUs  | `active_time_seconds` | `compute_time_seconds` |
   |--------|-----------------------|------------------------|
   | 0.25   | 1                     | 0.25                   |
   | 4      | 1                     | 4                      |

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+| Parameter (project)                 | Trial                     | Pro                      |
+|------------------------|---------------------------|--------------------------|
+| `active_time_seconds`  | 86,400 (1 day)            | 2,592,000 (30 days)      |
+| `compute_time_seconds` | 3,600 (1 hour)            | 50,000 (approx. 14 hours)|
```
If my assumptions about compute_time_seconds, I wonder if "1 hour" and "14 hours" makes sense. In our Free Tier, we allow a 1/4 vCPU to run 24/4, for example. So, the setting would be 1/4 of the active_time_seconds value.

bgrenon replied:
Does this look a bit better?


| Parameter (project)       | Trial  (.25 vCPU)               | Pro  (max 4 vCPU)                |
|---------------------------|---------------------------------|----------------------------------|
| `active_time_seconds`     | 633,600 (business month 22 days)| 2,592,000 (30 days)              |
| `compute_time_seconds`    | 158,400 (approx 44 hours)       | 10,368,000 (up to 4 times the active time)|
| `written_data_bytes`      | 1,000,000,000 (approx. 1 GiB)   | 50,000,000,000 (approx. 50 GiB)  |
| `data_transfer_bytes`     | 500,000,000 (approx. 500 MiB)   | 10,000,000,000 (approx. 10 GiB)  |

| Parameter (branch)        | Trial                           | Pro                              |
|---------------------------|---------------------------------|----------------------------------|
| `logical_size_bytes`      | 100,000,000 (approx. 100 MiB)   | 10,000,000,000 (approx. 10 GiB)  |

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+
+_**What happens when the quota is met?**_
+
+When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming proxy conne ...
```
Should we mention who would have to intervene? Is it Neon Support or the partner?

bgrenon replied:
It's on the partner. Revised to:

"If you don't take explicit action otherwise, the suspension remains in place until the end of the current billing period starts (`consumption_period_end`)."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,346 @@
+
+
+## Resetting a project after suspend
+Generally, projects remain suspended until the next billing period. It is good practice to notify your users when they are close to reaching a limit; if the user is then suspended and loses access to their database, it will not be unexpected. If you have configured no further actions, the user will have to wait until the next billing period starts to resume usage.
```
It says generally here. Are there exceptions?

bgrenon replied:
The exception -- I think -- is if the partner does something to reset. I'll remove "generally" though. It could be misleading.

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+
+The `quota` object includes an array of parameters used to set threshold limits. Their names generally match their corresponding metric:
+
+* `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated so far during the current billing period.
```
I don't think this part is true, since it elapsed time can include the time that your compute is active because there is an autosuspend delay period:
"This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated so far during the current billing period." 

I think it's just the total amount of time your project's compute endpoints are allowed to be active.

bgrenon replied:
Ah, right. The inactive period, before autosuspend kicks in, is counted. How about:

"Sets the maximum amount of time your project's total compute endpoints are allowed to be active during the current billing period. It excludes time when computes are in an `Idle` state due to [auto-suspension](/docs/reference/glossary#auto-suspend-compute)."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+updatedOn: '2023-11-09T17:36:14.275Z'
+---
+
+When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute size a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresho ...
```
```suggestion
When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute resources a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresholds after which all active computes for a project are [suspended](#suspending-active-computes). 
```

danieltprice commented on content/docs/guides/partner-billing.md:
````diff
@@ -0,0 +1,350 @@
+You can set quotas for these consumption metrics per project using the following Neon API URL:
+
+```bash
+/projects/{project_id}/settings/quota
````
I am wondering if providing this URL is useful.  Would it be more useful to say "using the quote object in the Create project or Update project API, and link to those methods?

bgrenon replied:
Revised to:

"You can set quotas for these consumption metrics per project using the `quote` object in the [Create project](https://api-docs.neon.tech/reference/createproject) or [Update project](https://api-docs.neon.tech/reference/updateproject) API."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+| Parameter (project)       | Trial  (.25 vCPU)               | Pro  (max 4 vCPU)                |
+|---------------------------|---------------------------------|----------------------------------|
+| active_time_seconds     | 633,600 (business month 22 days)| 2,592,000 (30 days)              |
+| compute_time_seconds    | 158,400 (approx 44 hours)       | 10,368,000 (up to 4 times the active time)|
```
```suggestion
| compute_time_seconds    | 158,400 (approx 44 hours)       | 10,368,000 (4 times the active time for 4 vCPUs)|
```

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+
+_**What happens when the quota is met?**_
+
+When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming proxy conne ...
```
```suggestion
When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming connection. If you don't take explicit action otherwise, the suspension remains in place until the end of the current billing period starts (`consumption_period_end`).
```

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+See [Querying metrics and quotas](#querying-metrics-and-quotas) to find your reset date, billing period, and other values related to the project's consumption.
+
+<Admonition type="Note">
+Neon tracks these consumption metrics on a monthly cycle. If you want to track metrics on a specific time range, you need to store your metrics as snapshots. You can also use the Preview [Consumption API](#retrieving-metrics-for-all-projects) to collect metrics from across a range of billing periods. 
```
```suggestion
Neon tracks these consumption metrics on a monthly cycle. If you want to track metrics on a different cycle, you need to take snapshots of your metrics at the desired interval and store the data externally. You can also use the Preview [Consumption API](#retrieving-metrics-for-all-projects) to collect metrics from across a range of billing periods. 
```

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+* [Update an existing project](#update-an-existing-project)
+
+### Set quotas when you create the project
+For performance reasons, you might want to configure these quotas at the same time that you create a new project for your user, reducing the number of API calls you need to make.
```
```suggestion
For performance reasons, you might want to configure these quotas at the same time that you create a new project for your user using the [Create a project](https://api-docs.neon.tech/reference/createproject) API, reducing the number of API calls you need to make.
```

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+### Set quotas when you create the project
+For performance reasons, you might want to configure these quotas at the same time that you create a new project for your user, reducing the number of API calls you need to make.
+
+Here is a sample `POST` in `curl` that creates a new project called `UserNew` and sets the `active_time_seconds` quota to a total allowed time of 10 hours (36,000 seconds) for the month, and a total allowed `compute_time_seconds` set to 2.5 hours (9,000 seconds) for the month.  
```
2.5 hours makes sense if the compute is a .25 CU compute. Should we mention or does it complicate matters?

bgrenon replied:
I think it's good to include. How about this as an additional sentence?

"This 4:1 ratio between active and compute time is suitable for a fixed compute size of 0.25 vCPU."

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+
+If you do not include these parameters, the query defaults to the current consumption period. 
+
+Here is an example query that returns metrics from September 1st and December 1st, 2023.
```
```suggestion
Here is an example query that returns metrics from September 1st and December 1st, 2023. Time values must be provided  in ISO 8601 format. You can use this [timestamp converter](https://www.timestamp-converter.com/).
```

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+If you do not include these parameters, the query defaults to the current consumption period. 
+
+Here is an example query that returns metrics from September 1st and December 1st, 2023.
+
```
<CodeBlock shouldWrap>

danieltprice replied:
Use `<CodeBlock shouldWrap>` and `</CodeBlock>` to wrap this long example.

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,350 @@
+* `cursor` &#8212; by default, the response uses the project `id` from the last project in the list as the `cursor` value (included in the `pagination` object at the end of the response). Generally, it is up to the application to collect and use this cursor value when setting up the next request.
+
+Here is an example `GET` request asking for the next 100 projects, starting with project id `divine-tree-77657175`:
+
```
Use `<CodeBlock shouldWrap>` and `</CodeBlock>` to wrap this long example.

danieltprice replied:
A question for Atli or someone who has worked with partners and "quotas": Should this document cover these project limits as well:

- autoscaling_limit_min_cu
- autoscaling_limit_max_cu
- suspend_timeout_seconds
- history_retention_seconds

Daniel-ef commented:
- BTW, should the quota limits be in the `For Partners` section? As it's available for all Pro users too.
- > Should this document cover these project limits as well:

  These limits don't relate to billing or consumption, so it should be somewhere else in the docs.
- In `quota` object we also have `logical_size_bytes`. It's not mentioned in `Manage billing with consumption limits` -> `Available metrics`, for example.

Actually didn't pay much attention to our docs. And surprisingly it's very detailed. My respect to #docs team!

acervantes23 commented:
@bgrenon this is really well done! Yes, let's cover:

autoscaling_limit_min_cu
autoscaling_limit_max_cu
suspend_timeout_seconds

For, the autoscaling_limit_max_cu you can mention that this sets the upper limit on how much they would pay for compute.

After that good to go for me.

bgrenon commented:
> * BTW, should the quota limits be in the `For Partners` section? As it's available for all Pro users too.
> * > Should this document cover these project limits as well:
>   
>   
>   These limits don't relate to billing or consumption, so it should be somewhere else in the docs.
> * In `quota` object we also have `logical_size_bytes`. It's not mentioned in `Manage billing with consumption limits` -> `Available metrics`, for example.
> 
> Actually didn't pay much attention to our docs. And surprisingly it's very detailed. My respect to #docs team!

bgrenon commented:
@danieltprice can you review these recent updates:

- Added `logical_size` to this section: [Available metrics](https://neon-next-git-bgrenon-partners-guide-neondatabase.vercel.app/docs/guides/partner-billing#available-metrics)
- Added new section for other settings:  [Other consumption related settings](https://neon-next-git-bgrenon-partners-guide-neondatabase.vercel.app/docs/guides/partner-billing#other-consumption-related-settings)

danieltprice commented on content/docs/guides/partner-billing.md:
```diff
@@ -0,0 +1,450 @@
+
+These consumption metrics represent total cumulative usage across all branches and computes in a given project, accrued so far in a given monthly billing period. Metrics are refreshed on the first day of the following month, when the new billing period starts. 
+
+There is an additional value you can set quota controls for: `logical_size`, which gives you the current size of a particular branch. 
```
So, Neon sets a branch size limit on Free and Pro (3GB and 200GB, respectively).

branch_logical_size_limit (MiB)
branch_logical_size_limit_bytes (Bytes)

Users can monitor their branch data size as it grows using the "Get branch" logical_size attribute.

Users can self-impose a branch size limit that applies to each branch individually using Create/Update project logical_size_bytes.

danieltprice replied:
And sorry for my confusion yesterday!

bgrenon replied:
This all makes sense - and it's also nice seeing all these topically related values laid out in one flow like this. It's clarifying.
